workflow/controller/retention/controller.go

```diff
@@ -0,0 +1,165 @@
+package retention
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	runtimeutil "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/argoproj/argo-workflows/v3/config"
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	workflow "github.com/argoproj/argo-workflows/v3/pkg/client/clientset/versioned"
+	"github.com/argoproj/argo-workflows/v3/util"
+	envutil "github.com/argoproj/argo-workflows/v3/util/env"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+	"github.com/argoproj/argo-workflows/v3/workflow/controller/indexes"
+	"github.com/argoproj/argo-workflows/v3/workflow/metrics"
+)
+
+type Interface interface {
+	Run(context.Context) error
+}
+
+var _ Interface = retention{}
+
+var retentionGCPeriod = envutil.LookupEnvDurationOr("RETENTION_GC_PERIOD", time.Minute)
+
+func New(f config.RetentionPolicy, i cache.SharedIndexInformer, wi workflow.Interface, metrics *metrics.Metrics) Interface {
+	return &retention{
+		policy:            f,
+		retentionGCPeriod: retentionGCPeriod,
+		wfInformer:        i,
+		workqueue:         metrics.RateLimiterWithBusyWorkers(workqueue.DefaultControllerRateLimiter(), "workflow_retention"),
+		workflowInterface: wi,
+	}
+}
+
+type retention struct {
+	policy            config.RetentionPolicy
+	retentionGCPeriod time.Duration
+	wfInformer        cache.SharedIndexInformer
+	workqueue         workqueue.RateLimitingInterface
+	workflowInterface workflow.Interface
+}
+
+func (r retention) Run(ctx context.Context) error {
+	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
+	defer r.workqueue.ShutDown()
+
+	if ok := cache.WaitForCacheSync(ctx.Done(), r.wfInformer.HasSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	go wait.Until(r.filter, r.retentionGCPeriod, ctx.Done())
+	go wait.Until(r.runWorker, time.Second, ctx.Done())
+
+	log.Info("Started retention worker")
+	<-ctx.Done()
+	log.Info("Shutting retention worker")
+	return nil
+}
+
+type count struct {
+	errored int
+	failed  int
+}
+
+func (r *retention) filter() {
+	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
+	objs, err := r.wfInformer.GetIndexer().ByIndex(indexes.WorkflowCompletedIndex, "true")
+	if err != nil {
+		log.Errorf("Failed to get by index: %v, error: %v", indexes.WorkflowCompletedIndex, err)
+		return
+	}
+	var uns []*unstructured.Unstructured
+	for _, obj := range objs {
```

Comment:
Does this work with 100k completed workflows?

Reply:
Could you elaborate your concerns here @alexec? I guess the only way to properly answer that question would be to actually do the load test.